src/Rector/AbstractScopeAwareRector.php

```diff
@@ -24,19 +21,18 @@ abstract class AbstractScopeAwareRector extends AbstractRector implements ScopeA
      */
     public function refactor(Node $node): int|array|Node|null
     {
-        /** @var MutatingScope|null $currentScope */
-        $currentScope = $node->getAttribute(AttributeKey::SCOPE);
-
-        if (! $currentScope instanceof Scope) {
-            $errorMessage = sprintf(
-                'Scope not available on "%s" node, but is required by a refactorWithScope() method of "%s" rule. Fix scope refresh on changed nodes first',
-                $node::class,
-                static::class,
-            );
+        trigger_error(
+            sprintf(
+                'The `Rector\Rector\AbstractScopeAwareRector` is removed, use `Rector\Rector\AbstractScopeAwareRector` instead, see upgrading guide %s',
+                'https://github.com/rectorphp/rector-src/blob/main/UPGRADING.md#1-abstractscopeawarerector-is-removed-use-abstractrector-instead',
+            ),
+            E_USER_DEPRECATED
+        );
 
-            throw new ShouldNotHappenException($errorMessage);
-        }
+        $currentScope = $node->getAttribute(AttributeKey::SCOPE);
 
+        // method called needs to trigger exception above
+        // as refactor() is called via refactorWithScope()
         return $this->refactorWithScope($node, $currentScope);
```

Comment:
Here it should fail with exception. Otherwise people won't upgrade and keep using this one.

Reply:
it can't, as `refactorWithScope()` method is created via user, and it is called `refactor()` so throw is never executed as override, if I define throw before, it unreachable, as too early, and user never call as well.